algorithms/src/std_algorithms/Kokkos_Constraints.hpp

```diff
@@ -0,0 +1,225 @@
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_STD_ALGORITHMS_CONSTRAINTS_HPP_
+#define KOKKOS_STD_ALGORITHMS_CONSTRAINTS_HPP_
+
+#include <Kokkos_View.hpp>
+
+namespace Kokkos {
+namespace Experimental {
+
+template <typename T, typename enable = void>
+struct is_admissible_to_kokkos_std_algorithms : std::false_type {};
+
+template <typename T>
+struct is_admissible_to_kokkos_std_algorithms<
+    T, std::enable_if_t< ::Kokkos::is_view<T>::value && T::rank == 1 &&
+                         (std::is_same<typename T::traits::array_layout,
+                                       Kokkos::LayoutLeft>::value ||
+                          std::is_same<typename T::traits::array_layout,
+                                       Kokkos::LayoutRight>::value ||
+                          std::is_same<typename T::traits::array_layout,
+                                       Kokkos::LayoutStride>::value)> >
+    : std::true_type {};
+
+template <class ViewType>
+KOKKOS_INLINE_FUNCTION constexpr void
+static_assert_is_admissible_to_kokkos_std_algorithms(const ViewType&) {
+  static_assert(is_admissible_to_kokkos_std_algorithms<ViewType>::value,
+                "Currently, Kokkos standard algorithms only accept 1D Views.");
+}
+
+//
+// is_iterator
+//
+template <class T, class = void>
+struct is_iterator : std::false_type {};
+
+template <class T>
+struct is_iterator<
+    T, std::enable_if_t<!std::is_void<typename T::iterator_category>::value> >
+    : std::true_type {};
```

Comment:
Use `is_detected` or `detected_t`.  Also I assume this is not the final version.

Reply:
is the detection idiom inside Kokkos somewhere? if so, I missed it

Reply:
https://github.com/kokkos/kokkos/blob/develop/core/src/Kokkos_DetectionIdiom.hpp

Reply:
ah! ok will use that! thanks for point it out